urwid/vterm.py

```diff
@@ -1036,21 +1037,44 @@ def erase(
 
             y += 1
 
-    def sgi_to_attrspec(self, attrs: Iterable[int], fg: int, bg: int, attributes: set[str]) -> AttrSpec | None:
+    def sgi_to_attrspec(self, attrs: Iterable[int], fg: int, bg: int, attributes: set[str], prev_colors: int) -> AttrSpec | None:
         """
         Parse SGI sequence and return an AttrSpec representing the sequence
         including all earlier sequences specified as 'fg', 'bg' and
         'attributes'.
         """
-        for attr in attrs:
+
+        idx = 0
+        colors = prev_colors
+
+        while idx < len(attrs):
+            attr = attrs[idx]
             if 30 <= attr <= 37:
                 fg = attr - 30
+                colors = max(16, colors)
             elif 40 <= attr <= 47:
                 bg = attr - 40
-            elif attr == 38:
-                # set default foreground color, set underline
-                attributes.add('underline')
-                fg = None
+                colors = max(16, colors)
+            elif attr == 38 or attr == 48:
+                if idx + 2 < len(attrs) and attrs[idx + 1] == 5:
+                    # 8 bit color specification
+                    color = attrs[idx + 2]
+                    colors = max(256, colors)
+                    if attr == 38:
+                        fg = color
+                    else:
+                        bg = color
+                    idx += 2
+                elif idx + 4 < len(attrs) and attrs[idx + 1] == 2:
+                    # 24 bit color specification
+                    color = (attrs[idx + 2] << 16) + \
+                        (attrs[idx + 3] << 8) + attrs[idx + 4]
+                    colors = 2**24
+                    if attr == 38:
+                        fg = color
+                    else:
+                        bg = color
+                    idx += 4
             elif attr == 39:
                 # set default foreground color, remove underline
                 attributes.discard('underline')
```

Comment:
`underline` was effect of previous attribute `38` processing, looks like need to remove this side-effect

Reply:
Done.